### PR TITLE
Re-install Softaculous after upgrade

### PR DESCRIPTION
**Pull request by sloanebernstein**
Case RE-846: It was reported that the Softaculous plugin for cPanel was displaying blank pages after ELevate, and that a re-install of the plugin fixed this issue. That seems reasonable, so detect Softaculous and re-install if present.

Changelog: Re-install Softaculous after upgrade.

By submitting pull requests to this repo, I agree to the Contributor License Agreement which can be found at: https://github.com/cpanel/elevate/blob/main/docs/cPanel-CLA.pdf

### CODE REVIEW

File: lib/Elevate/Components/Softaculous.pm [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        program => Cpanel::Binaries::path('php'),
+        args    => [ $SOFTACULOUS_CLI_PATH, '--version' ],
+    );
+
+    return if $sr->exec_failed() || $sr->to_exception();
+
+    my $version = $sr->stdout();
+    chomp $version;
+
+    if ($version) {
```
Comment (toddr) [inline lib/Elevate/Components/Softaculous.pm:49]: length $version

File: lib/Elevate/Components/Softaculous.pm [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+        INFO('Softaculous has been detected. The system will re-install that software after the distro upgrade.');
+        Elevate::StageFile::update_stage_file( { softaculous => $version } );
+    }
+
+    return;
+}
+
+sub post_distro_upgrade ($self) {
+
+    my $version = Elevate::StageFile::read_stage_file( 'softaculous', '' );
```
[the remark this answers is not in the document]
Comment (cPholloway) [reply lib/Elevate/Components/Softaculous.pm:59]: The second argument is the default value to return if the 1st argument is missing from the stagefile

File: lib/Elevate/Components/Softaculous.pm [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    return unless $version;
+
+    my $path = Elevate::Fetch::script( 'https://files.softaculous.com/install.sh', 'softaculous_install' );
+
+    if ($path) {
+        INFO('Re-installing Softaculous:');
+        LOGDIE('Re-installation of Softaculous failed.') if !$self->ssystem( $path, '--reinstall' );
```
Comment (cPholloway) [inline lib/Elevate/Components/Softaculous.pm:66]: I don't think that this failing should be fatal for the script.  Can we provide the customer with some information on how to reinstall softaculous to get it working again and add this to the final notifications at the end instead?

**Review by cPholloway — CHANGES_REQUESTED** [review]
Please add a unit test for the new component

File: lib/Elevate/Components/Softaculous.pm (hunk 3 above)
Comment (sloanebernstein) [reply lib/Elevate/Components/Softaculous.pm:66]: I went back and forth about whether it should be fatal or not. Changing it back.

Comment (toddr) [reply lib/Elevate/Components/Softaculous.pm:66]: Is it a problem they can clean up after Elevate completes? Continue. If it's going to break the rest of elevate then stop and make them fix it.

Comment (cPholloway) [reply lib/Elevate/Components/Softaculous.pm:66]: > Is it a problem they can clean up after Elevate completes?

Yes, this is what customers are having to do now.  The released version of Elevate completes successfully now without any of this in place.

**Review by cPholloway — DISMISSED** [review]
Approved pending follow up case to finish the unit test. Needs assertions for `post_distro_upgrade`

**Review by cPholloway — APPROVED** [review]

**Pull request merged**